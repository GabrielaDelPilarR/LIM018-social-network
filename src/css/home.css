.home {
    background: linear-gradient(168.8deg, #FFF5F3 -8.88%, #FFEFEF 102.78%);
    width: 100%;
    height: 100vh;
}

.headerHome {
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding-block: 0.5rem;
    padding-inline: 2rem;
}

.homeH2{
    margin: 0;
}

.headerInfo{
    padding-top: 0.5rem;
}

.publicacion {
    width: 300px;
    height: 79px;
    background: #FFFFFF;
    border: 1px solid #F8C7CF;
    box-shadow: 0px 6px 6px rgba(0, 0, 0, 0.25);
    border-radius: 5px;
    padding: 1rem;
    font: var(--fontMain);
}
.publicacion::placeholder{
    font: var(--fontMain);
    padding: 0.4rem;
}

.createpost{
    display: flex;
    justify-self: center;
    align-items: center;
    flex-direction: column;
    gap: 1rem;
    width: 300px;
    height: 79px;
    background: #FFFFFF;
    border: 1px solid #F8C7CF;
    box-shadow: 0px 6px 6px rgba(0, 0, 0, 0.25);
    border-radius: 5px;
    padding: 1rem;
    font: var(--fontMain);
    margin: auto;
}

.iconosPost{
   
    width: 100%;
    display: flex;
    justify-content: space-between;
}

.course {
    display: flex;
    margin: auto;
    justify-content: start;
    align-items: flex-end;
    justify-content: space-between;
    background-image: linear-gradient(180deg, rgba(0, 0, 0, 0) 0%, #000000 77%),url(../img/juane.jpg);
    background-position: center;
    background-size: contain;
    border-radius: 0.5rem;
    width: 327px;
    height: 200px;
    margin-block: 2rem;
    
}

.second {

    background-image: linear-gradient(180deg, rgba(0, 0, 0, 0) 0%, #000000 77%),url(../img/picante.jpg);
    background-position: center;
    background-size: contain;
    background-repeat: repeat;
    border-radius: 0.5rem;
}

.like {
    display: block;
    background-image: url(../img/corazon.png);
    background-repeat: no-repeat;
    background-position: center;
    width: 3rem;
    height: 3rem;
    padding: 0;
    background-color: #F63D5D;
    border-radius: 50%;
    margin: 1rem;
}

.like:focus {
    background-image: url(../img/corazonActivo.png);
    background-repeat: no-repeat;
    background-position: center;
}

h3 {
    font: var(--fontMain);
    font-weight: bold;
    color: #FFEFEF;
    margin: 0;
}

p {
    color: #FFEFEF;
    font: var(--fontMain);
    font-size: 11px;
    margin-block-start: 6px;
   
}

.contentPubli {
    padding-inline: 1rem;
}

.btnPublicar{
    display: block;
    color: #FFFFFF;
    width: 98px;
    height: 24px;
    padding-block: 0;
    padding-inline: 0.5rem;
    background: #F63D5D;
    border-radius: 4px;
}

.sectionSearch{
    display: flex;
    justify-content: space-between;
    justify-items: end;
    color:  #424141;;
    background-color: #FFEFEF;
    position: fixed;
    bottom: 0;
    width: 100%;
    align-items: flex-end;

}

.cajita{
   border: 1px solid red;
    padding: 1.2rem ;
    border-radius: 1rem 1rem 0 0 ;
    font-size: 14px;
    background-color: #FFEFEF;
    filter: drop-shadow(0px 4px 4px rgba(0, 0, 0, 0.25));
  

}