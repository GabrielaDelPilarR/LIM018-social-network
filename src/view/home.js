import {
  observer,
  savePost,
  onGetPost,
  deletePost,
  getPost,
  updatePost,
} from '../firebase/auth.js';

const home = {
  template: () => { /* Backticks: Permiten concatenar y trabajar con cadenas */
    const homeTemplate = `   
    <section class="home">
      <div class="header-home">
        <h1>HELP TASTER</h1>
        <div class="description-img">
          <img class="photo-user" src="" referrerpolicy="no-referrer">
        </div>
      </div>
      <div id = "contentPost" class = "content-post"></div>

      <div class="nav">
        <div class="home-nav">
          <button id="btn-home-nav" class="btn-nav">
            <img src="./img/recipe (Stroke).png">Home</button>
        </div>
        <div class="publicar-nav">
          <button id="btn-publicar-nav" class="btn-nav">
            <img src="./img/photo_camera.png">Publicar
          </button>
        </div>
        <div class="buscar-nav">
          <button id="btn-buscar-nav" class=" btn-nav">
            <img src="./img/search.png">Buscar</button>
        </div>
      </div>

      <div class="modal-container">
        <div class="modal no-verified-email">
          <form id="create-post" class="create-post">
            <label><input type="text" id = "name-restaurant" class="name-restaurant" placeholder="Restaurante"></label>
            <textarea id="publicacion" class="publicacion" placeholder="¿Por qué nos recomiendas este lugar?"></textarea>
              <div class="iconos-post">
                <div>
                  <span class="material-symbols-outlined">image</span>
                </div>
                <button id="btn-publicar" class="btn-primary btn-publicar">Publicar</button>
                <button class="btn-primary btn-close-modal">Cancelar</button>
              </div>
          </form>
        </div>
      </div>
    </section>
    `;

    const divRegister = document.createElement('div');
    divRegister.classList.add('registers');
    divRegister.innerHTML = homeTemplate;
    return divRegister;
  },

  init: () => {
    // Creando post
    const containerPost = document.querySelector('#contentPost');
    const postForm = document.querySelector('.create-post');
    const nameRestaurant = document.querySelector('#name-restaurant');
    const post = document.querySelector('.publicacion');
    const modalPublication = document.querySelector('.modal-container');
    let currentUser;

    /* let currentUser; */
    // Traer el nombre de usuario, (el observador)
    function authCallBack(user) {
      currentUser = user; // Usuario actual
      const photoUser = document.querySelector('.photo-user');
      photoUser.setAttribute('src', user.photoURL); // Cambia el contenido src x la foto
      if (user.photoURL == null) {
        photoUser.setAttribute('src', '../img/photo-user.png');
      }
    }
    observer(authCallBack);

    // funciones para bloquear y activar scroll del modal
    function blockScroll() {
      document.querySelector('.content-post').classList.add('hidden-scroll');
    }

    function activateScroll() {
      document.querySelector('.content-post').classList.remove('hidden-scroll');
    }

    // funcion para mostrar modal
    function showModal(configModal = {}) {
      const noopFunction = () => {};
      const btnCloseModal = document.querySelector('.btn-close-modal');

      const {
        continueText = 'Publicar',
        clickContinue = noopFunction,
        beforeLoad = noopFunction,
        onClose = noopFunction,
      } = configModal;

      postForm['btn-publicar'].innerText = continueText;

      postForm['btn-publicar'].addEventListener('click', clickContinue);
      btnCloseModal.addEventListener('click', onClose);

      beforeLoad();

      modalPublication.classList.add('show-modal-publication');
      // Esta en nav.css
    }

    // funcion para remover modal
    function removeModal(clickContinue = () => {}) {
      postForm['btn-publicar'].removeEventListener('click', clickContinue);

      modalPublication.classList.remove('show-modal-publication');
      activateScroll();
    }

    // Menú desplegable
    function menuPublication() {
      const menusDesplegables = document.querySelectorAll('.img-tree-dots');
      menusDesplegables.forEach((menuDesplegable) => {
        menuDesplegable.addEventListener('click', (event) => {
          const btnMenu = event.target.closest('.menu-desplegable').querySelector('.btn-edit-delete');
          if (btnMenu.classList.contains('show-menu')) {
            btnMenu.classList.remove('show-menu');
          } else {
            btnMenu.classList.add('show-menu');
          }
        });
      });
    }

    // Eliminando post
    function deletePostFinal() {
      const btnsDelete = containerPost.querySelectorAll('.btn-delete');
      btnsDelete.forEach((btn) => {
        btn.addEventListener('click', (event) => {
          deletePost(event.target.dataset.id);
        });
      });
    }

    // Editar post
    function editPostFinal() {
      const btnsEdit = containerPost.querySelectorAll('.btn-edit');

      btnsEdit.forEach((btn) => {
        btn.addEventListener('click', async (event) => {
          const editId = event.target.dataset.id;

          const clickContinue = (e) => {
            e.preventDefault();

            const content = post.value;
            updatePost(editId, { content });

            removeModal(clickContinue);
            postForm.reset();
          };

          showModal({
            continueText: 'Guardar',
            beforeLoad: async () => {
              const doc = await getPost(editId);
              const publication = doc.data();
              post.value = publication.content;
              blockScroll();
            },
            clickContinue,
            onClose: () => {
              removeModal(clickContinue);
            },
          });
        });
      });
    }
    function counterLike() {
      const btnLikes = document.querySelectorAll('.btn-likes');
      btnLikes.forEach((btnLike) => {
        btnLike.addEventListener('click', async (event) => {
          const elementBtnLike = event.target.closest('.btn-likes');
          const idPost = elementBtnLike.dataset.id;
          const doc = await getPost(idPost);
          const publication = doc.data();
          const likes = publication.likes + 1;
          updatePost(idPost, { likes });
        });
      });
    }

    // Haciendo el post
    onGetPost((querySnapshot) => { // Cuando ocurra 1 cambio te mando los nuevos dts
      let html = '';

      querySnapshot.forEach((doc) => {
        // Si el userID del post no es igual al id del currentUser no muestro el boton de eliminar
        const contentPost = doc.data();
        const avatarUser = contentPost.avatar === null ? './img/photo-user-blanco.png' : contentPost.avatar;
        /* console.log(contentPost.userID, currentUser.uid); */

        html += ` 
        <div class="container-publi">
          <div class="container-publi-img">

            ${contentPost.userID === currentUser.uid ? `<div class="menu-desplegable" >
              <button class="img-tree-dots" >
                <img src="../img/menu-desplegable.png">
              </button>
              <ul class="btn-edit-delete">
                <li><button type="button" class="btn-edit menu" data-id="${doc.id}">Editar</button></li>
                <li><button type="button" class="btn-delete menu" data-id="${doc.id}">Eliminar</button></li>
              </ul>
            </div>` : ''}

            <div class="content-publi">            
              <img class="photo-user-post" src="${avatarUser}" referrerpolicy="no-referrer">
              <div>
              <p class="user-publi">${contentPost.userName}</p>
              <span class="time-publi">hace 2 horas</span>
              </div>
            </div>
          </div>
          <div class="info-publi">
            <h4>${contentPost.title}</h4>
            <p class="description">${contentPost.content}</p>

            <div class="interacciones">
              <button class="btn-interaccion btn-likes" data-id="${doc.id}"> 
              <img src = "../img/corazon.png">
              <span class="conteo">${contentPost.likes}</span>
              </button>
              <button class="btn-interaccion">
              <img src = "../img/coment.png">
              <span class = "conteo comentarios">1234</span>
              </button>
            </div>
          </div>
        </div>
        `;
      });
      containerPost.innerHTML = html;
      menuPublication();
      deletePostFinal();
      editPostFinal();
      counterLike();
    });

    // evento cuando le dan click al boton del nav-publicar
    const btnPublicarNav = document.querySelector('#btn-publicar-nav');
    btnPublicarNav.addEventListener('click', () => {
      const clickContinue = (event) => {
        event.preventDefault();

        const userPublication = post.value;
        const userRestaurant = nameRestaurant.value;

        savePost({
          content: userPublication,
          title: userRestaurant,
          userName: currentUser.displayName,
          userID: currentUser.uid,
          avatar: currentUser.photoURL,
          urlImage: '',
          likes: 0,
          commets: [],
        });
        removeModal(clickContinue);
        postForm.reset();
      };

      showModal({
        beforeLoad: () => {
          blockScroll();
        },
        clickContinue,
        onClose: () => {
          removeModal(clickContinue);
        },
      });
    });
  },
};

export default home;
