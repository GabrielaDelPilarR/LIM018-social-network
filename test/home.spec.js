// import { getPost } from '../src/firebase/auth';
import home from '../src/view/home.js';

jest.mock('../src/firebase/auth.js', () => ({
  getPost: jest.fn(),
  observer: jest.fn().mockImplementationOnce((authCallBack) => authCallBack({ photoURL: null })),
  onGetPost: jest.fn(),
}));

describe('home', () => {
  it('should be a function', () => {
    expect(typeof home.template).toBe('function');
    expect(typeof home.init).toBe('function');
  });

  it('should check template', () => {
    expect(home.template()).toMatchSnapshot();
  });

  it('should render template home', () => {
    document.body.appendChild(home.template());
    home.init();
  });

  it('should open modal publish post', () => {
    document.body.appendChild(home.template());
    home.init();

    const btnPublish = document.querySelector('#btn-publicar-nav');
    btnPublish.click();
  });
});
// wt
