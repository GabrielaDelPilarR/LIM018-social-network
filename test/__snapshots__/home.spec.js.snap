// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home should check template 1`] = `
<div
  class="registers"
>
     
    
  <section
    class="home container-wrapper"
  >
    
      
    <div
      class="header-home"
    >
      
        
      <h1>
        HELP TASTER
      </h1>
      
        
      <div
        class="description-img"
      >
        
          
        <img
          class="photo-user"
          referrerpolicy="no-referrer"
          src=""
        />
        
          
        <div
          class="menu-salir"
        >
          
            
          <button
            class="img-salir"
          >
            
            
            <img
              src="../img/salir.png"
            />
            
            
          </button>
           
            
          <ul
            class="opciones-btn-salir"
          >
            
              
            <li>
              <button
                class="cerrar-sesion"
              >
                Cerrar sesión
              </button>
            </li>
            
              
            <li>
              <button
                class="cancelar"
              >
                Cancelar
              </button>
            </li>
            
            
          </ul>
          
          
        </div>
        
        
      </div>
      
      
    </div>
    

      
    <div
      class="wrapper"
    >
      
        
      <div
        class="container-publi container-perfil"
      >
        
          
        <div>
          
          
          <img
            class="photo-user-perfil"
            referrerpolicy="no-referrer"
            src=""
          />
          
          
        </div>
        
          
        <span>
          <strong
            id="currentName"
          />
        </span>
        
          
        <h3
          class="descripcion"
        >
          Descripción
        </h3>
        
          
        <span>
          Me gusta la cocina , soy  aficionad@. 
          La comida Peruana es mi favorita, ya que esta dentro de los primeros lugares gastronomicos del mundo.
          Quiero ampliar mi conocimiento respecto a los mejores restaurantes del Perú. 
          Mis platos  favoritos son el chaufa y el picante de camarones a la tacneña. Recomiendame tú lugar fovorito!! 
        </span>
        
          
        <div
          class="iconosFollow"
        >
          
            
          <img
            src="../img/group.png"
          />
          
            
          <span>
             23 followers  3 following 
          </span>
          
          
        </div>
        
        
      </div>
      

        
      <div
        class="content-post"
        id="contentPost"
      />
      
      
    </div>
    

      
    <div
      class="nav"
    >
      
        
      <div
        class="home-nav"
      >
        
          
        <button
          class="btn-nav"
          id="btn-home-nav"
        >
          
            
          <img
            src="./img/recipe (Stroke).png"
          />
          Home
        </button>
        
        
      </div>
      
        
      <div
        class="publicar-nav"
      >
        
          
        <button
          class="btn-nav"
          id="btn-publicar-nav"
        >
          
            
          <img
            src="./img/photo_camera.png"
          />
          Publicar
          
        </button>
        
        
      </div>
      
        
      <div
        class="buscar-nav"
      >
        
          
        <button
          class=" btn-nav"
          id="btn-buscar-nav"
        >
          
            
          <img
            src="./img/search.png"
          />
          Buscar
        </button>
        
        
      </div>
      
      
    </div>
    

      
    <div
      class="modal-container"
    >
      
        
      <div
        class="modal no-verified-email"
      >
        
          
        <form
          class="create-post"
          id="create-post"
        >
          
            
          <label>
            <input
              class="name-restaurant"
              id="name-restaurant"
              placeholder="Restaurante"
              type="text"
            />
          </label>
          
            
          <textarea
            class="publicacion"
            id="publicacion"
            placeholder="¿Por qué nos recomiendas este lugar?"
          />
          
              
          <div
            class="iconos-post"
          >
            
                
            <div>
              
                  
              <span
                class="material-symbols-outlined"
              >
                image
              </span>
              
                
            </div>
            
                
            <button
              class="btn-primary btn-publicar"
              id="btn-publicar"
            >
              Publicar
            </button>
            
                
            <button
              class="btn-primary btn-close-modal"
            >
              Cancelar
            </button>
            
              
          </div>
          
          
        </form>
        
        
      </div>
      
      
    </div>
    
    
  </section>
  
    
</div>
`;
